eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs

```diff
@@ -0,0 +1,312 @@
+    mkMemoBytes,
+    mkMemoized,
+  )
+import Cardano.Ledger.SafeHash (SafeToHash)
+import Cardano.Ledger.Shelley.Scripts (nativeMultiSigTag)
+import Cardano.Slotting.Slot (SlotNo (..))
+import Control.DeepSeq (NFData (..))
+import Data.ByteString.Lazy (fromStrict)
+import Data.ByteString.Short (fromShort)
+import Data.Sequence.Strict (StrictSeq (Empty, (:<|)))
+import Data.Set (Set, member)
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks (..))
+
+-- ================================================================
+-- A pair of optional SlotNo.
+
+-- | ValidityInterval is a half open interval. Closed on the bottom, open on the top.
+--   A SNothing on the bottom is negative infinity, and a SNothing on the top is positive infinity
+data ValidityInterval = ValidityInterval
+  { invalidBefore :: !(StrictMaybe SlotNo),
+    invalidHereafter :: !(StrictMaybe SlotNo)
+  }
+  deriving (Ord, Eq, Generic, Show, NoThunks, NFData)
+
+encodeVI :: ValidityInterval -> Encode ('Closed 'Dense) ValidityInterval
+encodeVI (ValidityInterval f t) = Rec ValidityInterval !> To f !> To t
+
+instance ToCBOR ValidityInterval where
+  toCBOR vi = encode (encodeVI vi)
+
+decodeVI :: Decode ('Closed 'Dense) ValidityInterval
+decodeVI = RecD ValidityInterval <! From <! From
+
+instance FromCBOR ValidityInterval where
+  fromCBOR = decode decodeVI
+
+-- ==================================================================
+
+data TimelockRaw era
+  = Signature !(KeyHash 'Witness (EraCrypto era))
+  | AllOf !(StrictSeq (Timelock era)) -- NOTE that Timelock and
+  | AnyOf !(StrictSeq (Timelock era)) -- TimelockRaw are mutually recursive.
+  | MOfN !Int !(StrictSeq (Timelock era))
+  | -- Note that the Int may be negative in which case (MOfN (-2) [..]) is always True
+    TimeStart !SlotNo -- The start time
+  | TimeExpire !SlotNo -- The time it expires
+  deriving (Eq, Generic, NFData)
+
+deriving instance Era era => NoThunks (TimelockRaw era)
+
+deriving instance HashAlgorithm (HASH (EraCrypto era)) => Show (TimelockRaw era)
+
+-- | This function deconstructs and then reconstructs the timelock script
+-- to prove the compiler that we can arbirarily switch out the eras as long
+-- as the cryptos for both eras are the same.
+translateTimelock ::
+  forall era1 era2.
+  ( Era era1,
+    Era era2,
+    EraCrypto era1 ~ EraCrypto era2
+  ) =>
+  Timelock era1 ->
+  Timelock era2
+translateTimelock (TimelockConstr (Memo tl bs)) =
+  let rewrap rtl = TimelockConstr $ mkMemoBytes rtl (fromStrict $ fromShort bs)
+   in case tl of
+        Signature s -> rewrap $ Signature s
+        AllOf l -> rewrap . AllOf $ translateTimelock <$> l
+        AnyOf l -> rewrap . AnyOf $ translateTimelock <$> l
+        MOfN n l -> rewrap $ MOfN n (translateTimelock <$> l)
+        TimeStart x -> rewrap $ TimeStart x
+        TimeExpire x -> rewrap $ TimeExpire x
+
+-- These coding choices are chosen so that a MultiSig script
+-- can be deserialised as a Timelock script
```

Comment:
is this supposed to be attached to the `ToCBOR` instance? I know you didn't write these... but there seems to be a lot of stray comments that are not as lovely at attached haddocks...